lib/rspec/matchers/built_in/raise_error.rb

```diff
@@ -128,19 +128,23 @@ def verify_message
           values_match?(@expected_message, @actual_error.message.to_s)
         end
 
-        def prevent_invalid_expectations
-          what_to_raise = if expecting_specific_exception? && @expected_message
-                            "`expect { }.not_to raise_error(SpecificErrorClass, message)`"
-                          elsif expecting_specific_exception?
-                            "`expect { }.not_to raise_error(SpecificErrorClass)`"
-                          elsif @expected_message
-                            "`expect { }.not_to raise_error(message)`"
-                          end
+        def warn_about_invalid_expectations
+          type = if expecting_specific_exception? && @expected_message
+                   "`expect { }.not_to raise_error(SpecificErrorClass, message)`"
+                 elsif expecting_specific_exception?
+                   "`expect { }.not_to raise_error(SpecificErrorClass)`"
+                 elsif @expected_message
+                   "`expect { }.not_to raise_error(message)`"
+                 end
 
-          return unless what_to_raise
+          return unless type
 
-          specific_class_error = ArgumentError.new("#{what_to_raise} is not valid, use `expect { }.not_to raise_error` (with no args) instead")
-          raise specific_class_error
+          expression = "Using `#{type}` risks false positives, since `#{type}` " \
+                       "won't match when Ruby raises a different error (e.g. " \
+                       "`NoMethodError`, `NameError` or `ArgumentError`). "
+          alternative = "Instead consider using `expect {}.not_to raise_error`. "
+
+          warn_about expression, alternative if RSpec::Expectations.configuration.warn_about_potential_false_positives?
```

Comment:
Seems like the `RSpec::Expectations.configuration.warn_about_potential_false_positives?` check could go on the first line of the method and avoid any work if it's `false`, right?